pyinterprod/interpro/signature.py

```diff
@@ -716,3 +715,128 @@ def update_citation(cur: cx_Oracle.Cursor, pmid: int) -> Optional[str]:
+
+        for pmid in params:
+            if pmid2pubid[pmid] not in used_citations_pubid:
+                delete_citation(cur, pmid)
+            else:
+                try:
+                    if ip_hash[int(pmid)] != litpub_hash[str(pmid)]:
+                        update_citation(cur, pmid)
+                except KeyError:
+                    logger.warning(f"Citation with Pubmed_id: '{pmid}' not found.")
+
+
+def update_citation(cur: cx_Oracle.Cursor, pmid: int):
+    cur.execute(
+        """
+        MERGE INTO INTERPRO.CITATION IC
+        USING (
+                SELECT DISTINCT
+                  C.EXTERNAL_ID AS EXTERNAL_ID, I.VOLUME AS VOLUME, I.ISSUE AS ISSUE,
+                  I.PUBYEAR AS YEAR, C.TITLE AS TITLE, C.PAGE_INFO AS RAWPAGES,
+                  J.MEDLINE_ABBREVIATION AS MEDLINE_JOURNAL, J.ISO_ABBREVIATION AS ISO_JOURNAL,
+                  A.AUTHORS AS AUTHORS, LOWER(REGEXP_REPLACE(U.URL, '(^[[:space:]]*|[[:space:]]*$)')) AS DOI_URL
+                FROM CDB.CITATIONS@LITPUB C
+                  LEFT OUTER JOIN CDB.JOURNAL_ISSUES@LITPUB I
+                    ON C.JOURNAL_ISSUE_ID = I.ID
+                  LEFT JOIN CDB.CV_JOURNALS@LITPUB J
+                    ON I.JOURNAL_ID = J.ID
+                  LEFT OUTER JOIN CDB.FULLTEXT_URL_MEDLINE@LITPUB U
+                    ON (
+                        C.EXTERNAL_ID = U.EXTERNAL_ID AND
+                        UPPER(U.SITE) = 'DOI'
+                    )
+                  LEFT OUTER JOIN CDB.AUTHORS@LITPUB A
+                    ON (
+                      C.ID = A.CITATION_ID AND
+                      A.HAS_SPECIAL_CHARS = 'N'
+                    )
+                ) L
+        ON (TO_CHAR(IC.PUBMED_ID) = L.EXTERNAL_ID)
+        WHEN MATCHED THEN
+          UPDATE SET IC.VOLUME = L.VOLUME,
+                     IC.ISSUE = L.ISSUE,
+                     IC.YEAR = L.YEAR,
+                     IC.TITLE = L.TITLE,
+                     IC.RAWPAGES = L.RAWPAGES,
+                     IC.MEDLINE_JOURNAL = L.MEDLINE_JOURNAL,
+                     IC.ISO_JOURNAL = L.ISO_JOURNAL,
+                     IC.AUTHORS = L.AUTHORS,
+                     IC.DOI_URL = L.DOI_URL
+          WHERE PUBMED_ID = :1
+        """, (pmid,)
+    )
+
+
+def delete_citation(cur: cx_Oracle.Cursor, pmid: int):
+    cur.execute(
+        """
+            DELETE FROM INTERPRO.CITATION
+            WHERE PUBMED_ID = :1
+        """, (str(pmid),)
+    )
+
+
+def _get_used_citations_pubid(cur: cx_Oracle.Cursor) -> set[str]:
+    cur.execute(
+        """
+            SELECT * FROM (
+                SELECT PUB_ID
+                FROM INTERPRO.ENTRY2PUB
+                UNION
+                SELECT PUB_ID FROM INTERPRO.SUPPLEMENTARY_REF
+                UNION
+                SELECT m.PUB_ID FROM INTERPRO.METHOD2PUB m
+            )
+        """
+    )
+    current_citations = [citations for citations, in cur.fetchall()]
+    return set(current_citations)
```

Comment:
- Indentation of SQL query
- No need to wrap the three subqueries
- No need to save the results in a list, then convert it to a set: we can directly create the set
- Variable name: `citations` -> `pub_id` (each element in the sequences returned by `cur.fetchall()` is a single publication ID; `citations` would be of a sequence of IDs)

```suggestion
    cur.execute(
        """
        SELECT PUB_ID FROM INTERPRO.ENTRY2PUB
        UNION
        SELECT PUB_ID FROM INTERPRO.SUPPLEMENTARY_REF
        UNION
        SELECT PUB_ID FROM INTERPRO.METHOD2PUB
        """
    )
    return {pub_id for pub_id, in cur.fetchall()}
```